crates/component/src/wrappers/http.rs

```diff
@@ -0,0 +1,353 @@
+                    response_out,
+                    Err(error_code("failed to get outgoing body handle")),
+                );
+                return;
+            };
+            let Ok(mut response_write) = response_body.write() else {
+                ResponseOutparam::set(
+                    response_out,
+                    Err(error_code("failed to get write handle to outgoing body")),
+                );
+                return;
+            };
+
+            // Set the response before writing the body. At this point, an error can't be returned.
+            ResponseOutparam::set(response_out, Ok(response));
+            response_write
+                .write_all(&body)
+                .expect("failed to write body stream");
+            drop(response_write);
+            OutgoingBody::finish(response_body, None).expect("failed to finish outgoing body");
+        }
+        (None, Some((stream, body))) => {
+            let Ok(response_body) = response.body() else {
+                ResponseOutparam::set(
+                    response_out,
+                    Err(error_code("failed to get outgoing body handle")),
+                );
+                return;
+            };
+            let Ok(response_write) = response_body.write() else {
+                ResponseOutparam::set(
+                    response_out,
+                    Err(error_code("failed to get write handle to outgoing body")),
+                );
+                return;
+            };
+
+            // Set the response before writing the body. At this point, an error can't be returned.
+            ResponseOutparam::set(response_out, Ok(response));
+            loop {
+                match response_write.blocking_splice(&stream, u64::MAX) {
+                    Ok(0) | Err(StreamError::Closed) => break,
+                    Ok(_) => continue,
+                    Err(StreamError::LastOperationFailed(_)) => {
+                        // TODO: log error
+                        return;
+                    }
+                }
+            }
+            // Drop the input stream, finish the incoming body
+            drop(stream);
+            if let Some(body) = body {
+                IncomingBody::finish(body);
+            }
+            // Drop the output stream, finish the outgoing body
+            drop(response_write);
+            OutgoingBody::finish(response_body, None).expect("failed to finish outgoing body");
+        }
+        (Some(_), Some(_)) => {
+            ResponseOutparam::set(
+                response_out,
+                Err(error_code(
+                    "cannot set both body and body stream in response",
+                )),
+            );
+        }
+        (None, None) => ResponseOutparam::set(response_out, Ok(response)),
+    }
+}
+
+/// Helper function to construct an internal server `ErrorCode` with an error message.
+fn error_code(e: impl ToString) -> ErrorCode {
+    ErrorCode::InternalError(Some(e.to_string()))
```

Comment:
I guess it's fine for the first iteration, but we should not classify all errors as internal